### DISCUSSION TITLE
How Can I Fix the Display Flickering when I use ClearScreen() or fillScreen() for recording of new screen data?.

### DISCUSSION THREAD

**Discussion by Knottis** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#361] [Q&A]

**Comment by mrcodetastic** [comment]
You need to enable double buffering and then flip across to the DMA 'back' buffer that has been updated in the background.

Look at the 'SmoothDoubleBuffer' example.

Or if you want your own buffer using FastLED, then look at the 'ChainedPanelsScreenBuffer' as well.

**Reply by Knottis** [reply]
**Thanks for your help** 

Have now inserted “**mxconfig.double_buff = true**” and “**dma_display->flipDMABuffer()**” then flickering disappeared from the display.

_Hope you can create a simple guide and program example for us beginners on how to get started easily and which step we should be taken care of._

_**Thanks again!**_
Regards, Knottis
(Sorry for my bad English)


### Program update Ver.1.1
//####################################################################################
#include <ESP32-HUB75-MatrixPanel-I2S-DMA.h>


// FONTS
#include <Fonts/FreeSansBold24pt7b.h>   


#define PANEL_RES_X 64  		// Number of pixels wide of each INDIVIDUAL panel module.
#define PANEL_RES_Y 64  		// Number of pixels tall of each INDIVIDUAL panel module.

#define NUM_ROWS 1                       	// Number of rows of chained INDIVIDUAL PANELS
#define NUM_COLS 3                       		// Number of INDIVIDUAL PANELS per ROW
#define PANEL_CHAIN NUM_ROWS * NUM_COLS  	// Total number of panels chained one to another

// Change this to your needs, for details on VirtualPanel pls read the PDF!
#define SERPENT true
#define TOPDOWN false


// placeholder for the matrix object
MatrixPanel_I2S_DMA *dma_display = nullptr;


**void setup()** {
  // Module configuration
  HUB75_I2S_CFG mxconfig(
    PANEL_RES_X,  // module width
    PANEL_RES_Y,  // module height
    PANEL_CHAIN   // chain length
  );

  //----------------------------------------------------------------------------------------------------
  **mxconfig.double_buff = true;**      // Turn ON double buffer - Use double buffer (twice amount of RAM required)
  //----------------------------------------------------------------------------------------------------

  // OK, now we can create our matrix object
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);
}


**void loop()** {
  // Clear Screen
  // Clear dma_display…
  dma_display->clearscreen();
  // let's adjust default brightness to about 195 ==>> 75%
  dma_display->setBrightness8(195); 

  // Time 
  sprintf(Buffer, "%02d", hour(local));
  TempString = Buffer;
  TempString += ":";
  sprintf(Buffer, "%02d", minute(local));
  TempString += Buffer;
  TempString += ":";
  sprintf(Buffer, "%02d", second(local));
  TempString += Buffer;
  //
  // Set Tekstfont
  dma_display->setFont(&FreeSansBold24pt7b);
  // Set Coler - RED
  dma_display->setTextColor(dma_display->color565(255, 255, 255)); 
  dma_display->setCursor(6, 40);
  //
  // Show the new Time-data on the display
  dma_display->print(TempString);
  //
  //----------------------------------------------------------------------------------------------------
  **dma_display->flipDMABuffer();**          // IMPORTANT -> SEND Pixel Buffer to DMA / Panel Output!
  //----------------------------------------------------------------------------------------------------
  delay(1000);
}

//####################################################################################

**Comment by AhsenDogan01** [comment]
I am having the same problem. but if I set the "mxconfig.double_buff" variable to true, it doesn't work at all and I can't get a picture on the screen.I tried the examples about it, but it still doesn't work.But if I set the value to false it works but this time it flickers as it appears in the video. How can i solve this problem ?

**Reply by mrcodetastic** [reply]
@AhsenDogan01 - I need more details on the esp32 device your using, type of panels, number of panels, code. Etc.

Impossible to debug otherwise

**Reply by mrcodetastic** [reply]
Debug output would be useful as well.

**Reply by AhsenDogan01** [reply]
//Microcontroller :ESP32S NodeMCU
//Panel : 1/8 scan , driver : MBI15124 , P10 16X32
//I use to 2xROWS 2xCOLS

#include "Arduino.h"
#include "ESP32-HUB75-MatrixPanel-I2S-DMA.h"
#include "ESP32-VirtualMatrixPanel-I2S-DMA.h"
#include "FreeMono9pt7b.h"

  // Panel configuration
  #define PANEL_RES_X 32 // Number of pixels wide of each INDIVIDUAL panel module. 
  #define PANEL_RES_Y 16 // Number of pixels tall of each INDIVIDUAL panel module.
   
  #define NUM_ROWS 2 // Number of rows of chained INDIVIDUAL PANELS
  #define NUM_COLS 2 // Number of INDIVIDUAL PANELS per ROW
      
  // Change this to your needs, for details please read the PDF in 
  // the 'ChainedPanels'example folder!
  #define SERPENT true
  #define TOPDOWN true

  int PANEL_CHAIN =  NUM_ROWS*NUM_COLS;
  
  // placeholder for the matrix object
  MatrixPanel_I2S_DMA *dma_display = nullptr;

  // placeholder for the virtual display object
  VirtualMatrixPanel  *virtualDisp = nullptr;
  
  /******************************************************************************
   * Setup!
   ******************************************************************************/
  void setup()
  {
      HUB75_I2S_CFG mxconfig(
                  PANEL_RES_X,   // module width
                  PANEL_RES_Y,   // module height
                  PANEL_CHAIN    // chain length
      );


    mxconfig.driver = HUB75_I2S_CFG::MBI5124;

    mxconfig.clkphase = true;
    mxconfig.double_buff = false; // It doesn't work if i do "true"

    dma_display = new MatrixPanel_I2S_DMA(mxconfig);

    dma_display->setBrightness8(255);    

    dma_display->begin();

    virtualDisp = new VirtualMatrixPanel((*dma_display),NUM_ROWS,NUM_COLS, PANEL_RES_X, PANEL_RES_Y, SERPENT, TOPDOWN);

    dma_display->fillScreen(dma_display->color565(255, 0, 0));
    delay(500);
    dma_display->fillScreen(dma_display->color565(0, 255, 0));
    delay(500);
    dma_display->fillScreen(dma_display->color565(0, 0, 255));
    delay(500);
    dma_display->clearScreen();

    virtualDisp->setTextWrap(false);
	  
  }

  void loop() 
  {
        dma_display->clearScreen();

        virtualDisp->setFont(&FreeMono9pt7b);
        virtualDisp->setTextSize(1);
        virtualDisp->setTextColor(virtualDisp->color565(0, 0, 255));
        virtualDisp->setCursor(0, 26);
        virtualDisp->print("HELLO");

        //dma_display->flipDMABuffer(); // Not Working
        
        delay(1000);
        
      
  }

**Comment by mrcodetastic** [comment]
```
//Panel : 1/8 scan , driver : MBI15124 , P10 16X32
```

Per the readme the library doesn't support 1/8 scan panels out of the box. Use the [One_Eight_1_8_ScanPanel](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/tree/master/examples/One_Eight_1_8_ScanPanel) scan example

**Reply by AhsenDogan01** [reply]
I tried this. but the result is still the same. When I set the "mxconfig.double_buff" variable to "true", no image appears on the screen.

**Comment by Knottis** [comment]
### My "To-Do" list when I use the HUB75-MatrixPanel-DMA
-----
_**Always use approved drivers Panel Chips**_

P2-128x64-32S-V40 LED 1515  			-> **FM6124**
P3-64x64-32S-2012-16B     (P3-20190103)	-> **RUC7258D**
https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA

**Controll the Wiring to the _ESP32-S3-WROOM-1_ with the "HUB75-MatrixPanel-DMA" Software Driver**

When you **Updates** the “Software Driver” remember to insert NEW value to **E _PIN_DEFAULT** in the file **esp32s3-default-pins.hpp**

I use **Pin-17** to  E _PIN_DEFAUL

**esp32s3-default-pins.hpp** 
---
C:\Users\XXYYZZ\Documents\Arduino\libraries\ESP32_HUB75_LED_MATRIX_PANEL_DMA_Display\src\platforms\esp32s3\esp32s3-default-pins.hpp 

#define 	R1 _PIN_DEFAULT      4		
#define 	G1 _PIN_DEFAULT      5		
#define 	B1 _PIN_DEFAULT      6
#define 	R2 _PIN_DEFAULT      7
#define 	G2 _PIN_DEFAULT      15
#define 	B2 _PIN_DEFAULT      16

#define 	A _PIN_DEFAULT        18
#define 	B _PIN_DEFAULT        8
#define 	C _PIN_DEFAULT        3
#define 	D _PIN_DEFAULT        42
**#define 	E _PIN_DEFAULT         17**
          
#define 	LAT _PIN_DEFAULT     40
#define 	OE _PIN_DEFAULT      2
#define 	CLK _PIN_DEFAULT     41
